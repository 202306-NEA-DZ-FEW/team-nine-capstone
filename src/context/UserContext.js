import React, { createContext, useContext, useEffect, useState } from "react";

import { auth } from "@/lib/firebase/controller";

const UserContext = createContext();

export function UserProvider({ children }) {
    const [user, setUser] = useState(null);

    useEffect(() => {
        const isLogged = auth.onAuthStateChanged((authUser) => {
            if (authUser) {
                // User is signed in
                setUser(authUser);
            } else {
                // User is signed out
                setUser(null);
            }
        });

        // Clean up the subscription
        return () => isLogged();
    }, []);

    return (
        <UserContext.Provider value={{ user, setUser }}>
            {children}
        </UserContext.Provider>
    );
}

export function useUser() {
    return useContext(UserContext);
}
