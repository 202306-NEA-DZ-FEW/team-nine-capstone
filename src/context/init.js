import React from "react";

function init() {
    return <div>user context and other stuff</div>;
}

export default init;
