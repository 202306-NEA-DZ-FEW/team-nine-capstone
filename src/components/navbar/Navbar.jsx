import { signOut } from "firebase/auth";
import Link from "next/link";
import { useRouter } from "next/router";
import { useTranslation } from "next-i18next";
import React, { useEffect, useState } from "react";
import { BsTranslate } from "react-icons/bs";
import {
    VscAccount,
    VscChevronDown,
    VscChevronUp,
    VscChromeClose,
    VscMenu,
} from "react-icons/vsc";

import { auth } from "@/lib/firebase/controller";

import { useUser } from "@/context/UserContext";

function Navbar() {
    const [hasScrolled, setHasScrolled] = useState(false);

    const [isMenuOpen, setIsMenuOpen] = useState(false);
    const [isLangDropdownOpen, setIsLangDropdownOpen] = useState(false);
    const [isAccountDropdownOpen, setIsAccountDropdownOpen] = useState(false);
    const { user } = useUser();
    const { t } = useTranslation("common");
    const router = useRouter();

    const toggleMenu = () => {
        setIsMenuOpen(!isMenuOpen);
    };

    const toggleLangDropdown = () => {
        setIsLangDropdownOpen(!isLangDropdownOpen);
        setIsAccountDropdownOpen(false);
    };

    const toggleAccountDropdown = () => {
        setIsAccountDropdownOpen(!isAccountDropdownOpen);
        setIsLangDropdownOpen(false);
    };

    const closeMenu = () => {
        setIsMenuOpen(false);
        setIsLangDropdownOpen(false);
        setIsAccountDropdownOpen(false);
    };

    function handleSignOut(e) {
        e.preventDefault();
        signOut(auth)
            .then(() => {
                router.push("/");
            })
            .catch((err) => {
                err.message;
            });
    }
    useEffect(() => {
        const handleScroll = () => {
            const scrollY = window.scrollY;
            if (scrollY > 0) {
                setHasScrolled(true);
            } else {
                setHasScrolled(false);
            }
        };
        window.addEventListener("scroll", handleScroll);

        return () => {
            window.removeEventListener("scroll", handleScroll);
        };
    }, []);
    return (
        <div
            className={` relative top-0 left-0 w-full h-auto z-50  ${
                hasScrolled
                    ? "bg-black/80 text-white "
                    : "bg-black/20 backdrop-blur-lg text-black shadow-lg"
            }`}
        >
            <div className='mx-auto px-4 sm:px-8 md:px-0'>
                <div className='flex items-center justify-between p-3'>
                    <Link
                        href='/'
                        className='text-2xl font-bold cursor-pointer'
                        onClick={closeMenu}
                    >
                        Logo
                    </Link>
                    {/*SM view*/}
                    <div className='absolute right-0 p-5 sm:hidden'>
                        <VscMenu
                            className='text-2xl cursor-pointer'
                            onClick={toggleMenu}
                        />
                    </div>
                    <div className='hidden sm:flex space-x-4'>
                        <Link
                            href='/about'
                            className='cursor-pointer flex hover:bg-amber-400 hover:rounded-lg p-2'
                            onClick={closeMenu}
                        >
                            {t("About")}
                        </Link>
                        <Link
                            href='/events'
                            className=' fcursor-pointer flex hover:bg-amber-400 hover:rounded-lg p-2'
                            onClick={closeMenu}
                        >
                            {t("Events")}
                        </Link>
                        <div className='relative w-38'>
                            <span
                                className='cursor-pointer flex  hover:bg-amber-400 hover:rounded-lg p-2'
                                onClick={toggleLangDropdown}
                            >
                                <BsTranslate className='mr-5' />
                                {t("Language")}
                                {isLangDropdownOpen ? (
                                    <VscChevronUp />
                                ) : (
                                    <VscChevronDown />
                                )}
                            </span>
                            {isLangDropdownOpen && (
                                <div
                                    className={`absolute top-full left-0 w-full p-2 border-grey-800 border-3 rounded-b-lg shadow-lg backdrop-blur-md mt-3 ${
                                        hasScrolled
                                            ? "bg-black/80 text-white"
                                            : "bg-black/20 backdrop-blur-lg text-black"
                                    }`}
                                >
                                    {router.query.id ? (
                                        <>
                                            <Link
                                                href={{
                                                    pathname: "",
                                                    query: { id: user.uid },
                                                }}
                                                locale='en'
                                                className='block text-sm  p-2 w-38 hover:bg-amber-400 rounded-xl'
                                            >
                                                English
                                            </Link>
                                            <Link
                                                href={{
                                                    pathname: "",
                                                    query: { id: user.uid },
                                                }}
                                                locale='ar'
                                                className='block text-sm p-2 w-38 hover:bg-amber-400 rounded-xl'
                                            >
                                                العربية
                                            </Link>
                                        </>
                                    ) : (
                                        <>
                                            <Link
                                                href=''
                                                locale='en'
                                                className='block text-sm text-white p-2 w-38 hover-bg-amber-200 rounded-xl'
                                            >
                                                English
                                            </Link>
                                            <Link
                                                href=''
                                                locale='ar'
                                                className='block text-sm text-white p-2 w-38 hover-bg-amber-200 rounded-xl'
                                            >
                                                العربية
                                            </Link>
                                        </>
                                    )}
                                </div>
                            )}
                        </div>
                        <div className='relative'>
                            <span
                                className='cursor-pointer flex hover:bg-amber-400 hover:rounded-lg p-2 items-center space-x-2'
                                onClick={toggleAccountDropdown}
                            >
                                <VscAccount className='mr-5' />
                                {user ? t("Account") : t("Join Us")}
                                {isAccountDropdownOpen ? (
                                    <VscChevronUp />
                                ) : (
                                    <VscChevronDown />
                                )}
                            </span>
                            {isAccountDropdownOpen && (
                                <div
                                    className={`absolute top-full left-0 w-full p-2 border-grey-800 border-3 rounded-b-lg shadow-lg mt-3 ${
                                        hasScrolled
                                            ? "bg-black/80 text-white"
                                            : "bg-black/20 backdrop-blur-lg text-black"
                                    }`}
                                >
                                    {user ? (
                                        <>
                                            <Link
                                                href={`/profile/${user.uid}`}
                                                className='block text-sm  p-2  hover:bg-amber-400 rounded-xl'
                                                onClick={closeMenu}
                                            >
                                                {t("Your Profile")}
                                            </Link>
                                            <Link
                                                href={`/profile/${user.uid}/events`}
                                                className='block text-sm  p-2  hover:bg-amber-400 rounded-xl'
                                                onClick={closeMenu}
                                            >
                                                {t("Your Events")}
                                            </Link>
                                            <Link
                                                href='/'
                                                className='block text-sm  p-2  hover:bg-amber-400 rounded-xl'
                                                onClick={handleSignOut}
                                            >
                                                {t("Sign Out")}
                                            </Link>
                                        </>
                                    ) : (
                                        <>
                                            <Link
                                                href='/authentication/signUp'
                                                className='block text-sm  p-2  hover:bg-amber-400 rounded-lg'
                                                onClick={closeMenu}
                                            >
                                                {t("Sign Up")}
                                            </Link>
                                            <Link
                                                href='/authentication/signIn'
                                                className='block text-sm  p-2  hover:bg-amber-400 rounded-lg'
                                                onClick={closeMenu}
                                            >
                                                {t("Sign In")}
                                            </Link>
                                        </>
                                    )}
                                </div>
                            )}
                        </div>
                    </div>

                    <div className='sm:hidden'>
                        {isMenuOpen && (
                            <div className='fixed top-0 left-0 w-screen h-auto bg-black/80 backdrop-blur-md text-white z-50'>
                                <div className='flex justify-between p-4'>
                                    <Link
                                        href='/'
                                        className='text-2xl font-bold cursor-pointer '
                                        onClick={closeMenu}
                                    >
                                        Logo
                                    </Link>

                                    <VscChromeClose
                                        className='text-2xl cursor-pointer'
                                        onClick={toggleMenu}
                                    />
                                </div>
                                <div className='text-center'>
                                    <Link
                                        href='/about'
                                        className='w-full p-2 flex items-center justify-center hover:bg-amber-400 hover:text-white border-t'
                                        onClick={closeMenu}
                                    >
                                        {t("About")}
                                    </Link>
                                    <Link
                                        href='/events'
                                        className='w-full p-2 flex items-center justify-center hover:bg-amber-400 hover:text-white  border-t'
                                        onClick={closeMenu}
                                    >
                                        {t("Events")}
                                    </Link>
                                    <div className='w-full border-t'>
                                        <button
                                            className='w-full p-2 flex items-center justify-center hover:bg-amber-400 hover:text-white '
                                            onClick={toggleLangDropdown}
                                        >
                                            <BsTranslate className='mr-5' />
                                            {t("Language")}{" "}
                                            {isLangDropdownOpen ? (
                                                <VscChevronUp />
                                            ) : (
                                                <VscChevronDown />
                                            )}
                                        </button>
                                        {isLangDropdownOpen && (
                                            <div className='w-full '>
                                                {router.query.id ? (
                                                    <>
                                                        <Link
                                                            href={{
                                                                pathname: "",
                                                                query: {
                                                                    id: user.uid,
                                                                },
                                                            }}
                                                            locale='en'
                                                            className='w-full p-2 flex items-center justify-center hover:bg-amber-300 hover:text-white rounded-lg'
                                                        >
                                                            English
                                                        </Link>

                                                        <Link
                                                            href={{
                                                                pathname: "",
                                                                query: {
                                                                    id: user.uid,
                                                                },
                                                            }}
                                                            locale='ar'
                                                            className='w-full p-2 flex items-center justify-center hover:bg-amber-300 hover:text-white rounded-lg'
                                                        >
                                                            العربية{" "}
                                                        </Link>
                                                    </>
                                                ) : (
                                                    <>
                                                        <Link
                                                            href=''
                                                            locale='en'
                                                            className='w-full p-2 flex items-center justify-center hover:bg-amber-400 hover:text-white 
                                                    '
                                                        >
                                                            English
                                                        </Link>
                                                        <Link
                                                            href=''
                                                            locale='ar'
                                                            className='w-full p-2 flex items-center justify-center hover:bg-amber-400 hover:text-white '
                                                        >
                                                            العربية{" "}
                                                        </Link>
                                                    </>
                                                )}
                                            </div>
                                        )}
                                    </div>
                                    donate
                                    <div className='relative'>
                                        <div
                                            className='w-full p-2 flex items-center justify-center hover:bg-amber-400 hover:text-white border-t'
                                            onClick={toggleAccountDropdown}
                                        >
                                            <VscAccount className='mr-4' />
                                            {user ? t("Account") : t("Join Us")}
                                            {isAccountDropdownOpen ? (
                                                <VscChevronUp />
                                            ) : (
                                                <VscChevronDown />
                                            )}
                                        </div>

                                        {isAccountDropdownOpen && (
                                            <div className='w-full border-'>
                                                {user ? (
                                                    <>
                                                        <Link
                                                            href={`/profile/${user.uid}`}
                                                        >
                                                            <button className='w-full p-2 flex items-center justify-center hover:bg-amber-400 hover:text-white '>
                                                                {t(
                                                                    "Your Profile"
                                                                )}
                                                            </button>
                                                        </Link>
                                                        <Link
                                                            href={`/profile/${user.uid}/events`}
                                                        >
                                                            <button className='w-full p-2 flex items-center justify-center hover:bg-amber-400 hover:text-white '>
                                                                {t(
                                                                    "Your Events"
                                                                )}
                                                            </button>
                                                        </Link>
                                                        <button
                                                            href='/'
                                                            className='w-full p-2 flex items-center justify-center hover:bg-amber-400 hover:text-white '
                                                            onClick={
                                                                handleSignOut
                                                            }
                                                        >
                                                            {t("Sign Out")}
                                                        </button>
                                                    </>
                                                ) : (
                                                    <>
                                                        <Link href='/authentication/signIn'>
                                                            <button className='w-full p-2 flex items-center justify-center hover:bg-amber-400 hover:text-white '>
                                                                {t("Sign In")}
                                                            </button>
                                                        </Link>
                                                        <Link href='/authentication/signUp'>
                                                            <button className='w-full p-2 flex items-center justify-center hover:bg-amber-400 hover:text-white '>
                                                                {t("Sign Up")}
                                                            </button>
                                                        </Link>
                                                    </>
                                                )}
                                            </div>
                                        )}
                                    </div>
                                </div>
                            </div>
                        )}
                    </div>
                </div>
            </div>
        </div>
    );
}

export default Navbar;
