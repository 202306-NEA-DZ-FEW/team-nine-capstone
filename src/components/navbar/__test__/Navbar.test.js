import renderer from "react-test-renderer";

import { useUser } from "@/context/UserContext";

import Navbar from "../Navbar";

jest.mock("src/context/UserContext.js");

// Mock the Firebase services
jest.mock("@/lib/firebase/controller.js", () => ({
    getAuth: jest.fn(),
    auth: { currentUser: jest.fn() },
}));

//Mock the Firestore collection and its methods
jest.mock("@/lib/firebase/controller.js", () => ({
    collection: jest.fn(),
}));

// Mock the Firestore instance
jest.mock("@/lib/firebase/controller.js", () => ({
    db: {
        collection: jest.fn(),
    },
}));

jest.mock("next/router", () => ({
    useRouter: () => ({
        push: jest.fn(),
    }),
}));

describe("Navbar", () => {
    it("renders correctly", () => {
        // Mock the useUser hook to provide a dummy implementation
        useUser.mockReturnValue({ user: { uid: "dummyUserId" } });

        const tree = renderer.create(<Navbar />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
