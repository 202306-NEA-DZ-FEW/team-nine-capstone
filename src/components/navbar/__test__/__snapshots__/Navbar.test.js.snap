// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders correctly 1`] = `
<div
  className="sticky left-0 top-0 w-full h-auto z-50  bg-bgc-silver text-txtc-DarkCharcoal shadow-sm font-Montserrat"
>
  <div
    className="mx-auto px-1 sm:px-2 md:px-0"
  >
    <div
      className="flex items-center justify-between p-1"
    >
      <a
        className="flex items-center"
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        <img
          alt="Logo"
          className="mr-2"
          data-nimg="1"
          decoding="async"
          height={35}
          loading="lazy"
          onError={[Function]}
          onLoad={[Function]}
          src="/_next/image?url=%2Fimages%2FnavLogo.png&w=96&q=75"
          srcSet="/_next/image?url=%2Fimages%2FnavLogo.png&w=48&q=75 1x, /_next/image?url=%2Fimages%2FnavLogo.png&w=96&q=75 2x"
          style={
            Object {
              "color": "transparent",
            }
          }
          width={35}
        />
        <span
          className="nav-logo"
        >
          webtitle
        </span>
      </a>
      <div
        className="absolute right-0 p-5 sm:hidden"
      >
        <svg
          className="text-2xl cursor-pointer"
          fill="currentColor"
          height="1em"
          onClick={[Function]}
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 5H0V4h16v1zm0 8H0v-1h16v1zm0-4.008H0V8h16v.992z"
          />
        </svg>
      </div>
      <div
        className="hidden sm:flex space-x-4"
      >
        <a
          className="cursor-pointer flex hover:bg-amber-400 hover:rounded-lg p-2"
          href="/about"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          About
        </a>
        <a
          className=" fcursor-pointer flex hover:bg-amber-400 hover:rounded-lg p-2"
          href="/events"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          Events
        </a>
        <div
          className="relative w-38"
        >
          <span
            className="cursor-pointer flex  hover:bg-amber-400 hover:rounded-lg p-2"
            onClick={[Function]}
          >
            <svg
              className="mr-5"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4.545 6.714 4.11 8H3l1.862-5h1.284L8 8H6.833l-.435-1.286H4.545zm1.634-.736L5.5 3.956h-.049l-.679 2.022H6.18z"
              />
              <path
                d="M0 2a2 2 0 0 1 2-2h7a2 2 0 0 1 2 2v3h3a2 2 0 0 1 2 2v7a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2v-3H2a2 2 0 0 1-2-2V2zm2-1a1 1 0 0 0-1 1v7a1 1 0 0 0 1 1h7a1 1 0 0 0 1-1V2a1 1 0 0 0-1-1H2zm7.138 9.995c.193.301.402.583.63.846-.748.575-1.673 1.001-2.768 1.292.178.217.451.635.555.867 1.125-.359 2.08-.844 2.886-1.494.777.665 1.739 1.165 2.93 1.472.133-.254.414-.673.629-.89-1.125-.253-2.057-.694-2.82-1.284.681-.747 1.222-1.651 1.621-2.757H14V8h-3v1.047h.765c-.318.844-.74 1.546-1.272 2.13a6.066 6.066 0 0 1-.415-.492 1.988 1.988 0 0 1-.94.31z"
              />
            </svg>
            Language
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clipRule="evenodd"
                d="M7.976 10.072l4.357-4.357.62.618L8.284 11h-.618L3 6.333l.619-.618 4.357 4.357z"
                fillRule="evenodd"
              />
            </svg>
          </span>
        </div>
        <div
          className="relative"
        >
          <span
            className="cursor-pointer flex hover:bg-amber-400 hover:rounded-lg p-2 items-center space-x-2"
            onClick={[Function]}
          >
            <svg
              className="mr-5"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16 7.992C16 3.58 12.416 0 8 0S0 3.58 0 7.992c0 2.43 1.104 4.62 2.832 6.09.016.016.032.016.032.032.144.112.288.224.448.336.08.048.144.111.224.175A7.98 7.98 0 0 0 8.016 16a7.98 7.98 0 0 0 4.48-1.375c.08-.048.144-.111.224-.16.144-.111.304-.223.448-.335.016-.016.032-.016.032-.032 1.696-1.487 2.8-3.676 2.8-6.106zm-8 7.001c-1.504 0-2.88-.48-4.016-1.279.016-.128.048-.255.08-.383a4.17 4.17 0 0 1 .416-.991c.176-.304.384-.576.64-.816.24-.24.528-.463.816-.639.304-.176.624-.304.976-.4A4.15 4.15 0 0 1 8 10.342a4.185 4.185 0 0 1 2.928 1.166c.368.368.656.8.864 1.295.112.288.192.592.24.911A7.03 7.03 0 0 1 8 14.993zm-2.448-7.4a2.49 2.49 0 0 1-.208-1.024c0-.351.064-.703.208-1.023.144-.32.336-.607.576-.847.24-.24.528-.431.848-.575.32-.144.672-.208 1.024-.208.368 0 .704.064 1.024.208.32.144.608.336.848.575.24.24.432.528.576.847.144.32.208.672.208 1.023 0 .368-.064.704-.208 1.023a2.84 2.84 0 0 1-.576.848 2.84 2.84 0 0 1-.848.575 2.715 2.715 0 0 1-2.064 0 2.84 2.84 0 0 1-.848-.575 2.526 2.526 0 0 1-.56-.848zm7.424 5.306c0-.032-.016-.048-.016-.08a5.22 5.22 0 0 0-.688-1.406 4.883 4.883 0 0 0-1.088-1.135 5.207 5.207 0 0 0-1.04-.608 2.82 2.82 0 0 0 .464-.383 4.2 4.2 0 0 0 .624-.784 3.624 3.624 0 0 0 .528-1.934 3.71 3.71 0 0 0-.288-1.47 3.799 3.799 0 0 0-.816-1.199 3.845 3.845 0 0 0-1.2-.8 3.72 3.72 0 0 0-1.472-.287 3.72 3.72 0 0 0-1.472.288 3.631 3.631 0 0 0-1.2.815 3.84 3.84 0 0 0-.8 1.199 3.71 3.71 0 0 0-.288 1.47c0 .352.048.688.144 1.007.096.336.224.64.4.927.16.288.384.544.624.784.144.144.304.271.48.383a5.12 5.12 0 0 0-1.04.624c-.416.32-.784.703-1.088 1.119a4.999 4.999 0 0 0-.688 1.406c-.016.032-.016.064-.016.08C1.776 11.636.992 9.91.992 7.992.992 4.14 4.144.991 8 .991s7.008 3.149 7.008 7.001a6.96 6.96 0 0 1-2.032 4.907z"
              />
            </svg>
            Account
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clipRule="evenodd"
                d="M7.976 10.072l4.357-4.357.62.618L8.284 11h-.618L3 6.333l.619-.618 4.357 4.357z"
                fillRule="evenodd"
              />
            </svg>
          </span>
        </div>
      </div>
      <div
        className="sm:hidden"
      />
    </div>
  </div>
</div>
`;
